File: c/lcd1602a.c
#include "lcd1602a.h"

void lcd1602a_init(struct lcd1602a_lcd* lcd) {
  lcd->gpio_setup(lcd->pinout->contrast);
  lcd->gpio_setup(lcd->pinout->data_command);
  lcd->gpio_setup(lcd->pinout->read_write);
  lcd->gpio_setup(lcd->pinout->clock);
  lcd->gpio_setup(lcd->pinout->d0);
  lcd->gpio_setup(lcd->pinout->d1);
  lcd->gpio_setup(lcd->pinout->d2);
  lcd->gpio_setup(lcd->pinout->d3);
  lcd->gpio_setup(lcd->pinout->d4);
  lcd->gpio_setup(lcd->pinout->d5);
  lcd->gpio_setup(lcd->pinout->d6);
  lcd->gpio_setup(lcd->pinout->d7);
  lcd->gpio_setup(lcd->pinout->led_plus);
  lcd->gpio_setup(lcd->pinout->led_minus);

  lcd1602a_send_command(lcd, LCD1602A_FUNCTION_SET);
  lcd1602a_send_command(lcd, LCD1602A_COMMAND_CLEAR);
  lcd1602a_send_command(lcd, 0b00000110); // Increment and shift cursor
  lcd1602a_send_command(lcd, 0b00001110); // Turn on display
  lcd1602a_send_command(lcd, LCD1602A_COMMAND_CLEAR);
}

void lcd1602a_print_char(struct lcd1602a_lcd* lcd, char c) {
  lcd1602a_send(lcd, c, LCD1602A_DATA_MODE);
}

void lcd1602a_print(struct lcd1602a_lcd* lcd, char* const text) {
  for (const char* c = text; *c != '\0'; c++) {
    lcd1602a_print_char(lcd, *c);
  }
}

void lcd1602a_send_command(struct lcd1602a_lcd* lcd, uint8_t command) {
  lcd1602a_send(lcd, command, LCD1602A_COMMAND_MODE);
}

void lcd1602a_send(struct lcd1602a_lcd* lcd, uint8_t value, LCD1602A_VALUE_T register_select) {
  lcd->gpio_set(lcd->pinout->data_command, register_select);
  lcd->gpio_set(lcd->pinout->read_write, LCD1602A_WRITE_MODE);

  lcd->gpio_set(lcd->pinout->d0, (value & 1)   ? 1 : 0);
  lcd->gpio_set(lcd->pinout->d1, (value & 2)   ? 1 : 0);
  lcd->gpio_set(lcd->pinout->d2, (value & 4)   ? 1 : 0);
  lcd->gpio_set(lcd->pinout->d3, (value & 8)   ? 1 : 0);
  lcd->gpio_set(lcd->pinout->d4, (value & 16)  ? 1 : 0);
  lcd->gpio_set(lcd->pinout->d5, (value & 32)  ? 1 : 0);
  lcd->gpio_set(lcd->pinout->d6, (value & 64)  ? 1 : 0);
  lcd->gpio_set(lcd->pinout->d7, (value & 128) ? 1 : 0);

  lcd->gpio_set(lcd->pinout->clock, 0);
  lcd->delay(1);
  lcd->gpio_set(lcd->pinout->clock, 1);
  lcd->delay(1);
  lcd->gpio_set(lcd->pinout->clock, 0);
}

File: python/lcd1602a.py
MODE_COMMAND = 0
MODE_DATA = 1
MODE_WRITE = 0
MODE_READ = 1

COMMAND_CLEAR = 0b00000001
COMMAND_RETURN_HOME = 0b00000010
COMMAND_NEWLINE = 0b11000000

OP_8_BIT = 0b00110000
OP_4_BIT = 0b00100000
OP_2_LINE = 0b00101000
OP_1_LINE = 0b00100000
OP_5_11_FONT = 0b00100100
OP_5_8_FONT = 0b00100000

CURSOR_OP_ON = 0b00001010
CURSOR_OP_POSITION_ON = 0b00001001


class Pins:
    def __init__(
        self,
        contrast: int,
        register_select: int,
        read_write: int,
        enable: int,
        data_pins: list[int],
        led_plus: int,
        led_minus: int,
    ):
        self.contrast = contrast
        self.register_select = register_select
        self.read_write = read_write
        self.enable = enable
        self.data_pins = data_pins
        self.led_plus = led_plus
        self.led_minus = led_minus


class Lcd:
    def __init__(self, pins: Pins, options: int, cursor_opts: int):
        self.pins = pins

        self._gpio_setup(self.pins.contrast)
        self._gpio_setup(self.pins.register_select)
        self._gpio_setup(self.pins.read_write)
        self._gpio_setup(self.pins.enable)
        for data_pin in self.pins.data_pins:
            self._gpio_setup(data_pin)
        self._gpio_setup(self.pins.led_plus)
        self._gpio_setup(self.pins.led_minus)

        self._send_command(options)
        self._send_command(0b00000110)
        self._send_command(0b00001100 | cursor_opts)
        self._send_command(COMMAND_CLEAR)

    def print_text(self, text: str):
        for char in text:
            if char == "\n":
                self._send_command(COMMAND_NEWLINE)
            else:
                self.print_char(char)

    def print_char(self, char: str):
        self._send(ord(char), MODE_DATA)

    def return_home(self):
        self._send_command(COMMAND_RETURN_HOME)

    def _gpio_set(self, pin: int, value: int):
        pass  # Placeholder for GPIO set implementation

    def _gpio_read(self, pin: int) -> int:
        pass  # Placeholder for GPIO read implementation

    def _gpio_setup(self, pin: int):
        pass  # Placeholder for GPIO setup implementation

    def _delay_ms(self, milliseconds: int):
        pass  # Placeholder for delay implementation

    def _send_command(self, command: int):
        self._send(command, MODE_COMMAND)

    def _send(self, value: int, register_select: int):
        self._gpio_set(self.pins.register_select, register_select)
        self._gpio_set(self.pins.read_write, MODE_WRITE)

        for index, data_pin in enumerate(self.pins.data_pins):
            bit_value = (value >> index) & 0x01
            self._gpio_set(data_pin, bit_value)

        self._gpio_set(self.pins.enable, 0)
        self._delay_ms(1)
        self._gpio_set(self.pins.enable, 1)
        self._delay_ms(1)
        self._gpio_set(self.pins.enable, 0)
